File: dsa/arrays/TwoSum.java
public class TwoSum {
    public int[] twoSum(int[] nums, int target) {
        java.util.HashMap<Integer, Integer> seen = new java.util.HashMap<>(nums.length * 2);
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            Integer index = seen.get(complement);
            if (index != null) {
                return new int[] {index, i};
            }
            seen.put(nums[i], i);
        }
        return new int[0];
    }
}

File: kotlin/arrays/TwoSum.kt
/**
 * LC 1 — Pattern: HashMap value → index
 * Pasted from phase1/Practice.kt
 */
class TwoSum {

    /** Brute force — check every pair. O(n²) time, O(1) space */
    fun twoSumBruteForce(nums: IntArray, target: Int): IntArray {
        for (i in nums.indices) {
            for (j in i + 1 until nums.size) {
                if (nums[i] + nums[j] == target) {
                    return intArrayOf(i, j)
                }
            }
        }
        return intArrayOf()
    }

    /** Best — one pass HashMap. O(n) time, O(n) space */
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val seen = hashMapOf<Int, Int>()
        for (i in nums.indices) {
            val need = target - nums[i]
            if (need in seen) {
                return intArrayOf(seen[need]!!, i)
            }
            seen[nums[i]] = i
        }
        return intArrayOf()
    }
}
